vermouth/map_parser.py

```diff
@@ -0,0 +1,403 @@
+        else:
+            current_block.merge_molecule(new_block)
+        return current_block
+
+    def add_block_from(self, block):
+        block = block.copy()
+        for node in block.nodes.values():
+            if 'replace' in node:
+                node.update(node['replace'])
+                del node['replace']
+        self.names.append(block.name)
+        self.blocks_from = self._add_block(self.blocks_from, block)
+
+    def add_block_to(self, block):
+        self.blocks_to = self._add_block(self.blocks_to, block)
+
+    def add_node_from(self, attrs):
+        if self.blocks_from is None:
+            self.blocks_from = Block()
+            idx = 1
+        else:
+            idx = max(self.blocks_from.nodes) + 1
+        self.blocks_from.add_node(idx, **attrs)
+    
+    def add_node_to(self, attrs):
+        if self.blocks_to is None:
+            self.blocks_to = Block()
+            idx = 1
+        else:
+            idx = max(self.blocks_to.nodes) + 1
+        self.blocks_to.add_node(idx, **attrs)
+
+    def add_edge_from(self, attrs1, attrs2):
+        nodes1 = list(self.blocks_from.find_atoms(**attrs1))
+        nodes2 = list(self.blocks_from.find_atoms(**attrs2))
+        assert len(nodes1) == len(nodes2) == 1
+        assert nodes1 != nodes2
+        self.blocks_from.add_edge(nodes1[0], nodes2[0])
+
+    def add_edge_to(self, attrs1, attrs2):
+        nodes1 = list(self.blocks_to.find_atoms(**attrs1))
+        nodes2 = list(self.blocks_to.find_atoms(**attrs2))
+        assert len(nodes1) == len(nodes2) == 1
+        assert nodes1 != nodes2
+        self.blocks_to.add_edge(nodes1[0], nodes2[0])
+
+    def add_mapping(self, attrs_from, attrs_to, weight):
+        nodes_from = list(self.blocks_from.find_atoms(**attrs_from))
+        nodes_to = list(self.blocks_to.find_atoms(**attrs_to))
+        assert len(nodes_from) == len(nodes_to) == 1
+        self.mapping[nodes_from[0]][nodes_to[0]] = weight
+
+    def add_reference(self, attrs_to, attrs_from):
+        nodes_to = list(self.blocks_to.find_atoms(**attrs_to))
+        assert len(nodes_to) == 1
+        node_to = nodes_to[0]
+        nodes_from = set(self.blocks_from.find_atoms(**attrs_from))
+        mapped_nodes = {from_ for from_ in self.mapping if node_to in self.mapping[from_]}
+        nodes_from = nodes_from.intersection(mapped_nodes)
+        assert len(nodes_from) == 1
+        self.references[node_to] = next(iter(nodes_from))
+
+    def get_mapping(self, map_type):
+        if self.blocks_from is None:
+            return None
+        mapping = Mapping(self.blocks_from, self.blocks_to, dict(self.mapping),
+                          self.references, ff_from=self.ff_from, ff_to=self.ff_to,
+                          type=map_type, names=tuple(self.names))
+        return mapping
+
+
+class MappingDirector:
+    RESNAME_NUM_SEP = '#'
+    RESIDUE_ATOM_SEP = ':'
+    MAP_TYPES = ['block', 'modification']
```

Comment:
Calling the section "block" is error prone. It looks like you are defining a block instead of a mapping. "mapping block" may be better?

"modification" is tricky as well. Indeed, as the force field and the mapping files now look so similar, we have to expect users mistaking one for the other. (We can also think about merging the parsers at some point to save on code duplication).

Reply:
It's not accidental that the fileformats resemble eachother. And merging parsers is a great idea, but for later.
I agree with the error prone-ness of the current headers.